src/services/cw721/cw721.service.ts

```diff
@@ -509,4 +517,182 @@ export default class Cw721HandlerService extends BullableService {
+      .first()
+      .throwIfNotFound();
+    // insert data from event_attribute_backup to event_attribute
+    const { limitRecordGet } = config.cw721.crawlMissingContract;
+    let done = false;
+    let currentId = 0;
+    while (!done) {
+      // eslint-disable-next-line no-await-in-loop
+      const histories = await SmartContractEvent.query()
+        .transacting(trx)
+        .alias('smart_contract_event')
+        .withGraphJoined(
+          '[message(selectMessage), tx(selectTransaction), attributes(selectAttribute), smart_contract(selectSmartContract).code(selectCode)]'
+        )
+        .modifiers({
+          selectCode(builder) {
+            builder.select('type');
+          },
+          selectTransaction(builder) {
+            builder.select('hash', 'height');
+          },
+          selectMessage(builder) {
+            builder.select('sender');
+          },
+          selectAttribute(builder) {
+            builder.select('key', 'value');
+          },
+          selectSmartContract(builder) {
+            builder.select('address', 'id');
+          },
+        })
+        .where('smart_contract:code.type', 'CW721')
+        .where('tx.height', '>', startBlock)
+        .andWhere('tx.height', '<=', endBlock)
+        .andWhere('smart_contract.id', smartContractId)
+        .andWhere('smart_contract_event.id', '>', currentId)
+        .orderBy('smart_contract_event.id', 'asc')
+        .limit(limitRecordGet);
+      if (histories.length === 0) {
+        done = true;
+        break;
+      }
+      const newHistories: CW721Activity[] = [];
+      // eslint-disable-next-line no-restricted-syntax
+      for (const history of histories) {
+        const onchainTokenId = getAttributeFrom(
+          history.attributes,
+          EventAttribute.ATTRIBUTE_KEY.TOKEN_ID
+        );
+        if (onchainTokenId) {
+          const token = (
+            await this.getCw721TokensRecordsWithTrx(
+              [
+                {
+                  contractAddress: history.smart_contract.address,
+                  onchainTokenId: getAttributeFrom(
+                    history.attributes,
+                    EventAttribute.ATTRIBUTE_KEY.TOKEN_ID
+                  ),
+                },
+              ],
+              trx
+            )
+          )[0];
+          if (token) {
+            const cw721TokenId = token.cw721_token_id;
+            newHistories.push(
+              CW721Activity.fromJson({
+                action: history.action,
+                sender: history.sender,
+                tx_hash: history.tx.hash,
+                cw721_contract_id: cw721Contract.id,
+                cw721_token_id: cw721TokenId,
+                height: history.tx.height,
+              })
+            );
+          }
+        }
+      }
+      if (newHistories.length > 0) {
+        // eslint-disable-next-line no-await-in-loop
+        await CW721Activity.query().insert(newHistories).transacting(trx);
```

Comment:
ra cả mảng rồi batchInsert sau

Reply:
do khả năng là newHistories khá to nên ko có biến chứa đc ạ